src/ir/module.h

```diff
@@ -104,20 +107,23 @@ class Block {
 // A class that contains a collection of blocks.
 class Module {
  public:
+  using BlockListType = std::vector<std::unique_ptr<Block>>;
+  Module() {}
+
   Module(const Module&) = delete;
   Module& operator=(const Module&) = delete;
   Module(Module&&) = default;
   Module& operator=(Module&&) = default;
 
   // Adds a block to the module and returns a pointer to it.
-  const Block* AddBlock(std::unique_ptr<Block> block) {
-    const Block* result = block.get();
+  const Block& AddBlock(std::unique_ptr<Block> block) {
```

Comment:
It might be more concise at call sites to just take the block by value (like `AddBlock(Block block)`) then use it to move-initialize the unique pointer (like `blocks_.push_back(std::make_unique<Block>(block))`).

Reply:
Isn't it more efficient just to pass in a `std::unique_ptr<>`?  Otherwise, we create something on the stack and then move it to the heap? Granted that most of the contents of `Block` is in the heap anyway and it may not be that much more expensive. However,  I am wondering if we really gain much in terms of conciseness by changing signature to  `AddBlock(Block block)`. WDYT?

Reply:
Not necessarily. If we create the `Block` as a temporary in the argument list to `AddBlock` like this:

```C++
const Block *canonical_block = module.AddBlock(Block(/*...*/));
```

Then the created `Block` will likely never touch the stack. Copy elision will cause it to be created directly into the parameter. If that parameter is `std::move`d, into the `std::unique_ptr`, that means that it will likely be created directly into the allocated memory.

Performance isn't my main concern here, I mostly just think it's more concise to, for some type `T`, pass in an argument of `T(/*params*/)` rather than `std::make_unique<T>(/*params*/)`. But if you like it the other way, I think it's fine as is.

Reply:
I see. We cannot initialize a `Block` with a single constructor call like `Block(...)`. e.g. An `Operation` in a block refers to the parent block. That is why we use `BlockBuilder` which returns `std::unique_ptr<Block>`.
  
Let us leave this for now and can revisit if it becomes a major pain point.